src/wp-includes/class-wp-block-metadata-registry.php

```diff
@@ -0,0 +1,237 @@
+ *
+ * @since 6.7.0
+ */
+class WP_Block_Metadata_Registry {
+
+	/**
+	 * Container for storing block metadata collections.
+	 *
+	 * Each entry maps a base path to its corresponding metadata and callback.
+	 *
+	 * @since 6.7.0
+	 * @var array<string, array<string, mixed>>
+	 */
+	private static $collections = array();
+
+	/**
+	 * Caches the last matched collection path for performance optimization.
+	 *
+	 * @since 6.7.0
+	 * @var string|null
+	 */
+	private static $last_matched_collection = null;
+
+	/**
+	 * Registers a block metadata collection.
+	 *
+	 * This method allows registering a collection of block metadata from a single
+	 * manifest file, improving performance for large sets of blocks.
+	 *
+	 * The manifest file should be a PHP file that returns an associative array, where
+	 * the keys are the block identifiers (without their namespace) and the values are
+	 * the corresponding block metadata arrays. The block identifiers must match the
+	 * parent directory name for the respective `block.json` file.
+	 *
+	 * Example manifest file structure:
+	 * ```
+	 * return array(
+	 *     'example-block' => array(
+	 *         'title' => 'Example Block',
+	 *         'category' => 'widgets',
+	 *         'icon' => 'smiley',
+	 *         // ... other block metadata
+	 *     ),
+	 *     'another-block' => array(
+	 *         'title' => 'Another Block',
+	 *         'category' => 'formatting',
+	 *         'icon' => 'star-filled',
+	 *         // ... other block metadata
+	 *     ),
+	 *     // ... more block metadata entries
+	 * );
+	 * ```
+	 *
+	 * @since 6.7.0
+	 *
+	 * @param string $path     The absolute base path for the collection ( e.g., WP_PLUGIN_DIR . '/my-plugin/blocks/' ).
+	 * @param string $manifest The absolute path to the manifest file containing the metadata collection.
+	 * @return bool True if the collection was registered successfully, false otherwise.
+	 */
+	public static function register_collection( $path, $manifest ) {
+		static $wpinc_path = '';
+		static $plugin_dir = '';
+
+		$path = wp_normalize_path( rtrim( $path, '/' ) );
+
+		if ( ! $wpinc_path ) {
+			$wpinc_path = wp_normalize_path( ABSPATH . WPINC );
+		}
+		if ( ! $plugin_dir ) {
+			$plugin_dir = wp_normalize_path( WP_PLUGIN_DIR );
+		}
+
+		// Check if the path is valid:
+		if ( str_starts_with( $path, $wpinc_path ) ) {
+			// Core path is valid.
+		} elseif ( str_starts_with( $path, $plugin_dir ) ) {
+			// For plugins, ensure the path is within a specific plugin directory and not the base plugin directory.
+			$relative_path = substr( $path, strlen( $plugin_dir ) + 1 );
+			$plugin_name   = strtok( $relative_path, '/' );
+
+			if ( empty( $plugin_name ) || $plugin_name === $relative_path ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block metadata collections can only be registered for a specific plugin. The provided path is neither a core path nor a valid plugin path.' ),
+					'6.7.0'
+				);
+				return false;
+			}
+		} else {
```

Comment:
The block registering API support registering blocks in theme directory. Both theme and child theme. Why is this not supported here?

Reply:
Good question. I originally made this to help out the plugins that were registering large collection of blocks, but didn't even consider that themes could be doing the same thing. I just now allowed the theme directory here: e93a3d16867, although I don't know of any such themes to check if they use the same directory convention.. ( `wp-content/themes/mytheme/blocks/myblockname/block.json` ).

Reply:
OK, I reverted the themes changes for now. Can re-examine with more time later.

Reply:
I disagree with @felixarntz here. I think we need to support theme resignation here.

Reply:
Based on what? I gave two good reasons why this doesn't have to be included. Please provide a rationale why you think this needs to be included.

Reply:
I opened https://core.trac.wordpress.org/ticket/62140 for us to discuss this further, to not unnecessarily block this PR.